### ISSUE TITLE
aux not being updated 

### ISSUE THREAD

**Issue by smarras79** [climate-machine/CLIMA#565]
The only way to update aux.SOMEMODEL.VAR and make it accessible from Diagnostics.jl is by assigning its value from `atmos_init_aux!`

When I am inside a model such as turbulence.jl or moisture.jl,  
the value of any auxiliary quantity `aux.SOMEMODEL.VAR` cannot be update and made available to Diagnostics.jl unless it is done from function `atmos_init_aux!`

Practical example:

In **turbulence.jl** we have the function:

```
function squared_buoyancy_correction(normS, ∇transform::Grad, aux::Vars)
  ∂θ∂Φ = dot(∇transform.turbulence.θ_v, aux.orientation.∇Φ)
  N² = ∂θ∂Φ / aux.moisture.θ_v
  Richardson = N² / (normS^2 + eps(normS))

  buoyancy_correction = sqrt(clamp(1 - Richardson*inv_Pr_turb, 0, 1))

  aux.turbulence.Ri = buoyancy_correction

  return buoyancy_correction
end
```

however, although we have `aux.turbulence.Ri = buoyancy_correction`, aux.turbulence.Ri is zero when shown in **Diagnostics.jl**

We should be able to update the value of `aux.SOMEMODE.VAR` from within any of the functions defined in SOMEMODEL.jl and not just from `atmos_init_aux!`.

How do we modify the code to make this possible?

(Above, SOMEMODEL could be turbulence, moisture, etc.)

@simonbyrne @akshaysridhar @charleskawczynski

**Comment by akshaysridhar** [comment]
@smarras79 : There is an `atmos_nodal_update_aux` function provided for updating auxiliary variables. For example, https://github.com/climate-machine/CLIMA/blob/c046b2f424c8ec59e8b42d76a93de97111940570/src/Atmos/Model/moisture.jl#L11 and https://github.com/climate-machine/CLIMA/blob/c046b2f424c8ec59e8b42d76a93de97111940570/src/Atmos/Model/moisture.jl#L51

**Comment by smarras79** [comment]
> @smarras79 : There is an `atmos_nodal_update_aux` function provided for updating auxiliary variables. For example,
> 
> https://github.com/climate-machine/CLIMA/blob/c046b2f424c8ec59e8b42d76a93de97111940570/src/Atmos/Model/moisture.jl#L11
> 
> and
> https://github.com/climate-machine/CLIMA/blob/c046b2f424c8ec59e8b42d76a93de97111940570/src/Atmos/Model/moisture.jl#L51
> 
> `atmos_init_aux!` does not update every timestep.

thanks @akshaysridhar; I see how it can work in the specific case that is used above, but it can only work if  `atmos_nodal_update_aux` has access to the quantities needed to calculate things such as the Richardson number or the dynamic SGS model.
What I am trying to convey is that it would be much more useful to allow any function to store whatever quantity we want into aux.

**Comment by slifer50** [comment]
@smarras79 @akshaysridhar I've done a relatively in depth following of the update_aux breadcrumb trail and I've concluded the following, please correct me if I'm wrong: 
Only the kernel is able to modify any of the MPIStateArrays as far as I know. Like what @smarras79 mentioned in his comment, update aux can't access the quantities it would need to calculate the information we want to extract, doing what @akshaysridhar  suggests is simple because we don't need anything that isn't already passed into atmos_nodal_update_aux.
The functions in turbulence.jl ( that provide the quantities we want to extract) are used through the kernel but only have access to a local temp aux array and thus can't modify it (I've tried to update an aux quantity through them to make sure of this) , thus nothing changes after initialization through init_aux. This presents two options so far as I can tell, in order to solve the issue: 1-allow the functions access through the kernel to the actual aux MPIStateArray and not a temp array, making it possible to modify aux through them (possibly dangerous depending on how we do it). This is probably the most efficient way since we only need make it possible for the functions to save to aux what they already calculate while the solve is running. Or 2-make it so that kernel_nodal_update_aux has access to the quantities necessary to then use the turbulence.jl functions and update aux accordingly.( This option adds a comparatively high amount of computational overhead since we would be recalculating gradients we already calculate elsewhere, but do not save.)

**Comment by blallen** [comment]
I ran into this issue and added the feature to access the diffusive state during `nodal_update_aux!` in order to save derivatives to the auxiliary state. Are you guys able to add all the gradients you want to save to the diffusive state? That way you only need to calculate them once. The downside is you are storing the same variable multiple times.

**Comment by slifer50** [comment]
@blallen Thanks for mentioning this, I'll look into trying that out.

**Comment by akshaysridhar** [comment]
@slifer50 , here's one way of setting that up(the update aux function has access to the diffuse vars - see the kernel file as well, similar to what Brandon suggested I think) `https://github.com/akshaysridhar/CLIMA/blob/ApplyAuxDiffusive/src/Atmos/Model/AtmosModel.jl`

**Comment by slifer50** [comment]
@blallen I've looked into saving things into the diffusive_state, however, all quantities saved to the diffusive state are penalized when calculating face visc terms, I'm not sure how to appropriately deal with this in a safe way, have you run into this problem?

**Comment by jkozdon** [comment]
> I ran into this issue and added the feature to access the diffusive state during `nodal_update_aux!` in order to save derivatives to the auxiliary state. Are you guys able to add all the gradients you want to save to the diffusive state? That way you only need to calculate them once. The downside is you are storing the same variable multiple times.

If these only need occasionally, a better approach might be to just write some custom kernels to compute them as needed. They should be fairly straightforward modifications of the existing viscous kernels.

> @blallen I've looked into saving things into the diffusive_state, however, all quantities saved to the diffusive state are penalized when calculating face visc terms, I'm not sure how to appropriately deal with this in a safe way, have you run into this problem?

If you computing something based on the gradient, I'd imagine you probably want to DG gradient that includes the penalization of the jumps. Also if you want to use values that are consistent with what's used elsewhere in the code (e.g., store something that is computed as part of the flux) you would want to do this.

That said, I'm not 100% sure what you all are trying to do so I could be totally wrong.

**Comment by slifer50** [comment]
@jkozdon I see what you mean, since they are gradients they should be penalized anyways. I'm currently looking to store gradients in the diffusive_state, I'm assuming the same penalty is used for gradients and for the diffusive_state. Please let me know if I'm mistaken.

**Comment by slifer50** [comment]
@jkozdon We are looking to try and see the turbulence models are behaving correctly by looking at some the quantities turbulence.jl calculates, these however need gradients and currently nodal_update_aux does not have access to any gradients, since diffusive! does have access to the gradients we want to first store them in the diffusive state then use them to calculate the quantities we then save to the auxstate

**Comment by blallen** [comment]
Hi @smarras79 and @slifer50, I implemented an `update_aux_diffusive!` function in #617. It should do everything you need.

**Comment by smarras79** [comment]
Thanks so much @blallen
